File: 221. Maximal Square.py
'''
https://leetcode.com/problems/maximal-square/
Given an m x n binary matrix filled with 0's and 1's, find the largest square containing only 1's and return its area.

Example 1:
Input: matrix = [["1","0","1","0","0"],["1","0","1","1","1"],["1","1","1","1","1"],["1","0","0","1","0"]]
Output: 4
'''
'''
Time:O(mn)
Space:O(mn)
'''
class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        s = 0
        dp = [[0]*(len(matrix[0])+1)for _ in range(len(matrix)+1)] 
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                if matrix[i][j]=='1':
                    dp[i+1][j+1]=min(dp[i][j+1], dp[i+1][j], dp[i][j])+1
                    s = max(s, dp[i+1][j+1])
        return s*s
    
class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        m, n = len(matrix), len(matrix[0])
        dp = [[0 for _ in range(n)]for _ in range(m)]
        res = 0
        for i in range(m):
            dp[i][0] = int(matrix[i][0])
            res = max(res, dp[i][0])
        for j in range(1, n):
            dp[0][j] = int(matrix[0][j])
            res = max(res, dp[0][j])
        for i in range(1, m):
            for j in range(1, n):
                if matrix[i][j]=='1':
                    dp[i][j] = min(dp[i-1][j-1], dp[i-1][j], dp[i][j-1])+1  
                    res = max(res, dp[i][j])
        print(dp)
        return res*res

File: 221. Maximal Square.java
/*
https://leetcode.com/problems/maximal-square/
Given an m x n binary matrix filled with 0's and 1's, find the largest square containing only 1's and return its area.

Example 1:
Input: matrix = [["1","0","1","0","0"],["1","0","1","1","1"],["1","1","1","1","1"],["1","0","0","1","0"]]
Output: 4
*/
/*
Time:O(mn)
Space:O(n)
*/
class Solution {
    public int maximalSquare(char[][] matrix) {
        int rows = matrix.length, cols = rows>0? matrix[0].length:0;
        int[] dp = new int[cols+1];
        int max = 0, prev = 0;
        for(int i=1; i<rows+1; i++){
            for(int j=1; j<cols+1; j++){
                int temp = dp[j];
                if(matrix[i-1][j-1]=='1'){
                    dp[j] = Math.min(Math.min(dp[j-1], dp[j]), prev)+1;
                    max = Math.max(max, dp[j]);
                }
                else    dp[j]=0;
                prev = temp;
            }
        }
        return max*max;
    }
}

/*
Time:O(mn)
Space:O(mn)
*/
class Solution {
    public int maximalSquare(char[][] matrix) {
        int rows = matrix.length, cols = rows>0? matrix[0].length:0;
        int[][] dp = new int[rows+1][cols+1];
        int max = 0;
        for (int i=1; i<rows+1; i++){
            for (int j=1; j<cols+1; j++){
                if (matrix[i-1][j-1]=='1'){
                    dp[i][j] = Math.min(Math.min(dp[i-1][j], dp[i][j-1]), dp[i-1][j-1])+1;
                    max = Math.max(max, dp[i][j]);
                }
            }
        }
        return max*max;
    }
}
